internal/query/projection/project_member.go

```diff
@@ -0,0 +1,131 @@
+package projection
+
+import (
+	"context"
+
+	"github.com/caos/logging"
+
+	"github.com/caos/zitadel/internal/errors"
+	"github.com/caos/zitadel/internal/eventstore"
+	"github.com/caos/zitadel/internal/eventstore/handler"
+	"github.com/caos/zitadel/internal/eventstore/handler/crdb"
+	"github.com/caos/zitadel/internal/repository/member"
+	"github.com/caos/zitadel/internal/repository/project"
+)
+
+type ProjectMemberProjection struct {
+	crdb.StatementHandler
+}
+
+const (
+	ProjectMemberProjectionTable = "zitadel.projections.project_members"
+)
+
+func NewProjectMemberProjection(ctx context.Context, config crdb.StatementHandlerConfig) *ProjectMemberProjection {
+	p := &ProjectMemberProjection{}
+	config.ProjectionName = ProjectMemberProjectionTable
+	config.Reducers = p.reducers()
+	p.StatementHandler = crdb.NewStatementHandler(ctx, config)
+	return p
+}
+
+func (p *ProjectMemberProjection) reducers() []handler.AggregateReducer {
+	return []handler.AggregateReducer{
+		{
+			Aggregate: project.AggregateType,
+			EventRedusers: []handler.EventReducer{
+				{
+					Event:  project.MemberAddedType,
+					Reduce: p.reduceAdded,
+				},
+				{
+					Event:  project.GrantMemberAddedType,
+					Reduce: p.reduceAdded,
+				},
+				{
+					Event:  project.MemberChangedType,
+					Reduce: p.reduceChanged,
+				},
+				{
+					Event:  project.MemberCascadeRemovedType,
+					Reduce: p.reduceCascadeRemoved,
+				},
+				{
+					Event:  project.MemberRemovedType,
+					Reduce: p.reduceRemoved,
+				},
+			},
+		},
+	}
+}
+
+type ProjectMemberColumn string
+
+const (
+	ProjectMemberProjectIDCol = "project_id"
+	ProjectMemberGrantIDCol   = "grant_id"
+)
+
+func (p *ProjectMemberProjection) reduceAdded(event eventstore.EventReader) (*handler.Statement, error) {
+	switch e := event.(type) {
+	case *project.MemberAddedEvent:
+		return reduceMemberAdded(e.MemberAddedEvent, withCol(ProjectMemberProjectIDCol, e.Aggregate().ID))
+	case *project.GrantMemberAddedEvent:
+		return reduceMemberAdded(
+			*member.NewMemberAddedEvent(&e.BaseEvent, e.UserID, e.Roles...),
+			withCol(ProjectMemberProjectIDCol, e.Aggregate().ID),
+			withCol(ProjectMemberGrantIDCol, e.GrantID),
+		)
+	default:
+		logging.LogWithFields("HANDL-tPdUI", "seq", event.Sequence(), "expectedTypes", []eventstore.EventType{project.MemberAddedType, project.GrantMemberAddedType}).Error("wrong event type")
+		return nil, errors.ThrowInvalidArgument(nil, "HANDL-rddM1", "reduce.wrong.event.type")
+	}
+}
+
+func (p *ProjectMemberProjection) reduceChanged(event eventstore.EventReader) (*handler.Statement, error) {
+	switch e := event.(type) {
+	case *project.MemberChangedEvent:
+		return reduceMemberChanged(e.MemberChangedEvent, withCond(ProjectMemberProjectIDCol, e.Aggregate().ID))
```

Comment:
triggers also project grant members?

Reply:
yes it's filled into the same table. If it's a grant member the grantID filter is added (case below)

Reply:
you're right, I decided to split the projection into project_member and project_grant_member